### PR TITLE
Implement Meta Get functionality

### PR DESCRIPTION
**Pull request by mrattle**
### TL;DR - What are you trying to accomplish?

This PR introduces the implementation of the `meta_get` method, which uses the meta parser added in [a previous PR](https://github.com/Shopify/async-memcached/pull/66).  There are also small fixes to parsing behaviour that became necessary as edge cases were revealed with actual use of the parser.

### Details - How are you making this change?  What are the effects of this change?

We are building the meta protocol in `async-memcached` with the full scope of flags available in mind.  As such, we are keeping methods broad, and extracting desired functionality from the general methods will be done through the flags provided and is left up to the user.  There is no prescribing of flags in this implementation.

Meta Get covers the [range of operations](https://github.com/memcached/memcached/blob/5609673ed29db98a377749fab469fe80777de8fd/doc/protocol.txt#L494) that correspond to `get`, `gets`, `gat`, `gats` and `touch` which all use different flag combinations.  `meta_get` takes optional flags because it is possible to bump the LRU for an item through what is essentially a no-op.  The `v` flag for value return does not need to be provided for all uses of `mg`, such as when using `touch` behaviour to update the TTL of a cached item.

================

Benchmarks for `meta_get` vs `meta_get_concat` with small key and value ("foo" & "bar"):

```
     Running benches/bench.rs (target/release/deps/bench-62483f1fc831355c)
Gnuplot not found, using plotters backend
get_small               time:   [33.272 µs 33.687 µs 34.316 µs]
                        change: [-8.7210% -7.6317% -6.4967%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) high mild
  3 (3.00%) high severe

bench meta_get          time:   [28.778 µs 28.873 µs 28.977 µs]
                        change: [-1.3716% -0.8567% -0.3263%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild
  8 (8.00%) high severe

bench meta_get_concat   time:   [29.076 µs 29.172 µs 29.284 µs]
                        change: [-1.0130% -0.5647% -0.1286%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe
```

Benchmarks for ASCII get vs Meta Get methods with max key size and max value size (250 bytes & 1MB):

```
     Running benches/bench.rs (target/release/deps/bench-62483f1fc831355c)
Gnuplot not found, using plotters backend
bench ascii get         time:   [245.16 µs 246.59 µs 248.90 µs]
                        change: [-0.5506% +1.6244% +4.9168%] (p = 0.29 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  7 (7.00%) high severe

bench meta_get          time:   [117.66 µs 117.89 µs 118.08 µs]
                        change: [+1.2427% +1.7885% +2.3200%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  6 (6.00%) high severe

bench meta_get_concat   time:   [117.76 µs 118.04 µs 118.33 µs]
                        change: [+1.0628% +1.7042% +2.3663%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low severe
  2 (2.00%) high mild
  4 (4.00%) high severe
```

TL;DR - `meta_get` and `meta_get_concat` are both faster than ASCII get under similar circumstances (ASCII get is ~16.2% slower in the small key/value test and ~2x slower in the worst case with max length key and value).  No statistical difference between the two `meta_get` methods in this style of benchmark at this scale, similar to what we have seen with the ASCII get methods.

### CODE REVIEW

**Comment by mrattle** [comment]
This branch is based off of the one that introduces the parsing functionality, so 75% of the commits will drop from here when that merges.

File: src/lib.rs [hunk 1]
```diff
@@ -214,6 +217,50 @@ impl Client {
...
+    /// Otherwise, `None` is returned.
+    ///
+    /// Supported meta flags:
+    /// - h: return whether item has been hit before as a 0 or 1
+    /// - l: return time since item was last accessed in seconds
+    /// - t: return item TTL remaining in seconds (-1 for unlimited)
+    pub async fn meta_get<K: AsRef<[u8]>>(
+        &mut self,
+        key: K,
+        meta_flags: Option<&[&str]>,
```
Comment (mrattle) [inline src/lib.rs:233]: `meta_flags` param here is a single optional slice of `&str`.  Could be swapped for `&[u8]` but I think the `&str` is fine in this case.  `Option<_>` is used because it's possible to perform a no-op on a key (providing no flags) to bump the LRU if desired.

One thing to consider is that this parameter would also include opaques.  It may be desirable to break the opaque token out as its own parameter.

File: src/parser/meta.rs [hunk 2]
```diff
@@ -1,35 +1,20 @@
...
+    sequence::tuple,
     IResult,
 };
 
+use std::num::NonZero;
+
 use super::{parse_u32, ErrorKind, MetaValue, Response, Status, Value};
 use crate::Error;
 
-// TODO: remove this later in favour of individual parse_meta_*_status methods
```
Comment (mrattle) [inline src/parser/meta.rs:18]: Removed this because using one general function doesn't work in practice, as expected.  There is some overlap between response codes in terms of differing interpretations and each op needs to be able to trigger the right logic when encountering a response code.

File: src/parser/meta.rs [hunk 3]
```diff
@@ -47,17 +32,27 @@ pub fn parse_meta_get_status(buf: &[u8]) -> IResult<&[u8], Response> {
...
-pub fn parse_meta_response(buf: &[u8]) -> Result<Option<(usize, Response)>, ErrorKind> {
+pub fn parse_meta_set_response(buf: &[u8]) -> Result<Option<(usize, Response)>, ErrorKind> {
     let bufn = buf.len();
-    // TODO: alt calls parsers sequentially until one succeeds, which is not optimal.  Want to only call the correct parser, no sequencing.
-    // Also run the risk of hitting the wrong parser since there is overlap in response codes.
-    let result = alt((
-        parse_meta_status,
-        |input| parse_meta_get_data_value(input),
-        |input| parse_meta_set_data_value(input),
-    ))(buf);
```
Comment (mrattle) [inline src/parser/meta.rs:60]: Removed because `alt` goes through unnecessary checks vs directly calling the intended method, and can also lead to unintended parsing behaviour (i.e. there are situations where a `meta_set` op could trigger `parse_meta_get_data_value` which has different logic & causes confusing behaviour)

File: src/parser/meta.rs [hunk 4]
```diff
@@ -233,23 +235,23 @@ fn parse_meta_set_data_value(buf: &[u8]) -> IResult<&[u8], Response> {
...
-        }),
-    ))(input)
+    // Slice the buffer to extract the data
+    let (extracted, remaining) = buf.split_at(size);
+
+    // Ensure the remaining buffer starts with "\r\n"
+    let (remaining, _) = tag("\r\n")(remaining)?;
+
+    Ok((remaining, extracted))
 }
```
Comment (mrattle) [inline src/parser/meta.rs:255]: Method no longer needed since everything is being parsed as a slice now.

File: src/parser/meta.rs [hunk 5]
```diff
@@ -598,7 +600,7 @@ mod tests {
         let input = b"HD\r\n";
         let (remaining, response) = parse_meta_get_data_value(input).unwrap();
 
-        assert_eq!(remaining, b"\r\n");
+        assert_eq!(remaining, b"");
```
Comment (mrattle) [inline src/parser/meta.rs:603]: Tests now confirm that the buffer is left empty instead of having a remaining `crlf` at the end of parsing.

**Review by danmayer — APPROVED** [review]
thanks this looks good to me, nice cleanup around the trailing newlines and testing that to stay consistent.

File: src/lib.rs [hunk 6]
```diff
@@ -214,6 +217,50 @@ impl Client {
...
+        self.conn.write_all(b"\r\n").await?;
+        self.conn.flush().await?;
+
+        match self.drive_receive(parse_meta_get_response).await? {
+            Response::Status(Status::NotFound) => Ok(None),
+            Response::Status(s) => Err(s.into()),
+            Response::Data(d) => d
+                .map(|mut items| {
+                    if items.len() != 1 {
+                        Err(Status::Error(ErrorKind::Protocol(None)).into())
```
Comment (grcooper) [inline src/lib.rs:252]: Do we have a more verbose error that we can surface here? What is this error case?

File: src/parser/meta.rs [hunk 7]
```diff
@@ -47,17 +32,27 @@ pub fn parse_meta_get_status(buf: &[u8]) -> IResult<&[u8], Response> {
...
 }
 
-// TODO: refactor this into individual parse_meta_*_response methods
+pub fn parse_meta_get_response(buf: &[u8]) -> Result<Option<(usize, Response)>, ErrorKind> {
+    let bufn = buf.len();
+    let result = parse_meta_get_data_value(buf);
+
+    match result {
+        Ok((left, response)) => {
+            let n = bufn - left.len();
```
Comment (grcooper) [inline src/parser/meta.rs:41]: Can we make these variables more explicit? It's hard to tell what is going on here.

File: src/lib.rs (hunk 6 above)
Comment (mrattle) [reply src/lib.rs:252]: I think this can actually just drop out.  It was carried over from the ASCII implementation but it doesn't actually make sense in this context.

In the ASCII protocol, theoretically if there is no key validation for whitespace in the client layer that uses this crate, it's possible to construct a `get` operation with either a key split by whitespace or multiple keys, which could yield > 1 result when the `get` API is meant to either fetch a single result or nothing.  In the ASCII protocol, `get_multi` is just a `get` with many keys, and the memcached server understands that syntax.  This is not the case in the meta protocol - `mg` can only be pipelined, so there can only ever be one result or `None`.

File: src/lib.rs [hunk 8]
```diff
@@ -214,6 +217,50 @@ impl Client {
...
+    /// - l: return time since item was last accessed in seconds
+    /// - t: return item TTL remaining in seconds (-1 for unlimited)
+    pub async fn meta_get<K: AsRef<[u8]>>(
+        &mut self,
+        key: K,
+        meta_flags: Option<&[&str]>,
+    ) -> Result<Option<Value>, Error> {
+        self.conn.write_all(b"mg ").await?;
+        self.conn
+            .write_all(Self::validate_key_length(key.as_ref())?)
```
Comment (nickamorim) [inline src/lib.rs:237]: Should we validate the key length _before_ we write `mg` into the buffer? We should fail as early as possible if the request is invalid

File: tests/integration_tests.rs [hunk 9]
```diff
@@ -84,6 +84,191 @@ async fn test_get_fails_with_key_too_long() {
...
+        Some("9001".as_bytes().to_vec())
+    );
+}
+
+#[ignore = "Relies on a running memcached server"]
+#[tokio::test]
+#[parallel]
+async fn test_meta_get_not_found_with_k_flag() {
+    let key = "meta-get-test-key-not-found";
+    let flags = ["v", "O9001"];
```
Comment (nickamorim) [inline tests/integration_tests.rs:264]: Also, why aren't we passing in a `k` flag as mentioned in the test name?

Comment (nickamorim) [inline tests/integration_tests.rs:264]: Why aren't we passing in a `k` flag as mentioned in the test name?

File: tests/integration_tests.rs [hunk 10]
```diff
@@ -2,11 +2,11 @@ use async_memcached::{Client, Error, ErrorKind, Status};
 use rand::seq::IteratorRandom;
 use serial_test::{parallel, serial};
 
-// Note: Each test should run with keys unique to that test to avoid async conflicts.  Because these tests run concurrently,
+// NOTE: Each test should run with keys unique to that test to avoid async conflicts.  Because these tests run concurrently,
 // it's possible to delete/overwrite keys created by another test before they're read.
 
-const LARGE_PAYLOAD_SIZE: usize = 1000 * 1024;
-const MAX_KEY_LENGTH: usize = 250;
+const LARGE_PAYLOAD_SIZE: usize = 1000 * 1024; // 1 MB, default memcached max value size
```
Comment (nickamorim) [inline tests/integration_tests.rs:8]: It's technically 1MiB (1024*1024)

File: tests/integration_tests.rs [hunk 11]
```diff
@@ -84,6 +84,191 @@ async fn test_get_fails_with_key_too_long() {
...
+async fn test_meta_get_cache_hit_with_no_flags() {
+    let key = "meta-get-test-key-cache-hit-with-no-flags";
+    let value = "test-value";
+
+    let mut client = setup_client(&[key]).await;
+
+    client.set(key, value, None, None).await.unwrap();
+
+    let get_result = client.get(key).await.unwrap();
+    println!(
```
Comment (nickamorim) [inline tests/integration_tests.rs:99]: Do we need this print?

**Review by nickamorim — COMMENTED** [review]
I'd like to see a test with the quiet flag passed in

File: src/parser/meta.rs (hunk 7 above)
Comment (mrattle) [reply src/parser/meta.rs]: Changed some names a little bit which hopefully makes it a bit more clear, but the code path is unfortunately a bit obscure by nature.

File: tests/integration_tests.rs (hunk 10 above)
Comment (nickamorim) [reply tests/integration_tests.rs]: The entire request has to be <1MiB so we use 1024*1024 - 250b (maximum key length) - 60b (metadata).

File: src/parser/meta.rs [hunk 12]
```diff
@@ -44,25 +30,35 @@ pub fn parse_meta_get_status(buf: &[u8]) -> IResult<&[u8], Response> {
...
-    let bufn = buf.len();
-    // TODO: alt calls parsers sequentially until one succeeds, which is not optimal.  Want to only call the correct parser, no sequencing.
-    // Also run the risk of hitting the wrong parser since there is overlap in response codes.
-    let result = alt((
-        parse_meta_status,
-        |input| parse_meta_get_data_value(input),
-        |input| parse_meta_set_data_value(input),
-    ))(buf);
+pub fn parse_meta_set_response(buf: &[u8]) -> Result<Option<(usize, Response)>, ErrorKind> {
+    let n_bytes_required = buf.len();
```
Comment (nickamorim) [inline src/parser/meta.rs:55]:
```suggestion
    let total_bytes = buf.len();
```

**Review by grcooper — APPROVED** [review]
